databricks_cli/stack/api.py

```diff
@@ -0,0 +1,340 @@
+        for deployed_resource in stack_status[STACK_DEPLOYED]:
+            if RESOURCE_ID not in deployed_resource:
+                raise StackError("%s doesn't exist in deployed resource status" % RESOURCE_ID)
+            if RESOURCE_SERVICE not in deployed_resource:
+                raise StackError("%s doesn't exist in deployed resource status" % RESOURCE_SERVICE)
+            if RESOURCE_PHYSICAL_ID not in deployed_resource:
+                raise StackError("%s doesn't exist in deployed resource status" %
+                                 RESOURCE_PHYSICAL_ID)
+
+    def validate_config(self, stack_config):
+        """
+        Validate fields within a stack configuration. This ensures that an inputted configuration
+        has the necessary fields for stack deployment to function well.
+
+        :param stack_config: dict- stack config that is inputted by the user.
+        :return: None. Raises errors to stop deployment if there is a problem.
+        """
+        if STACK_NAME not in stack_config:
+            raise StackError("'%s' not in configuration" % STACK_NAME)
+        if STACK_RESOURCES not in stack_config:
+            raise StackError("'%s' not in configuration" % STACK_RESOURCES)
+        for resource in stack_config[STACK_RESOURCES]:
+            if RESOURCE_ID not in resource:
+                raise StackError("%s doesn't exist in resource config" % RESOURCE_ID)
+            if RESOURCE_SERVICE not in resource:
+                raise StackError("%s doesn't exist in resource config" % RESOURCE_SERVICE)
+            if RESOURCE_PROPERTIES not in resource:
+                raise StackError("%s doesn't exist in resource config" % RESOURCE_PROPERTIES)
+
+    def deploy_config(self, stack_config, stack_status=None):
+        self.validate_config(stack_config)
+        if stack_status:
+            self.validate_status(stack_status)
+            resource_to_deployed_resource = self._get_previous_resource_deploy_map(stack_status)
+        else:
+            resource_to_deployed_resource = {}
+
+        stack_name = stack_config[STACK_NAME]
+        click.echo('Deploying stack %s' % stack_name)
+        deployed_resources = []
+
+        for resource in stack_config[STACK_RESOURCES]:
+            click.echo()
+            click.echo("Deploying resource")
+            # Retrieve resource deployment info from the last deployment.
+            resource_map_key = (resource[RESOURCE_ID], resource[RESOURCE_SERVICE])
+            resource_deploy_info = resource_to_deployed_resource[resource_map_key] \
+                if resource_map_key in resource_to_deployed_resource else None
+            # Deploy resource, get resource_status
+            resource_status = self.deploy_resource(resource, resource_deploy_info)
+            deployed_resources.append(resource_status)
+
+        # stack deploy status is original config with deployed resource statuses added
+        new_stack_status = stack_config
```

Comment:
Does this do a deep-copy? If not, do a deep-copy instead of mutating the `stack_config`.